internal/providers/registry.go

```diff
@@ -1,52 +1,318 @@
+		for {
+			r.actions <- func() {
+				doneChan <- r.pendingDisco == 0
+			}
+			done := <-doneChan
+			if done {
+				break
+			}
+			time.Sleep(200 * time.Millisecond)
+		}
+
+		close(r.actions)
+
+		if r.dstore != nil {
+			err = r.dstore.Close()
+		}
+	})
+	<-r.closed
+	return err
+}
+
+func (r *Registry) run() {
+	defer close(r.closed)
+
+	for action := range r.actions {
+		action()
+	}
+}
+
+func (r *Registry) Discover(fcAddr string, signature, signed []byte, sync bool) error {
+	errCh := make(chan error, 1)
+	r.actions <- func() {
+		r.syncStartDiscover(fcAddr, signature, signed, errCh)
+	}
+	if sync {
+		return <-errCh
+	}
+	return nil
+}
+
+func (r *Registry) Register(info *ProviderInfo) error {
+	if len(info.AddrInfo.ID) == 0 {
+		return errors.New("missing peer id")
+	}
+
+	// If provider is trusted, register immediately
+	if !r.policy.Trusted(info.AddrInfo.ID) {
+		return errors.New("provider must be trusted to register without on-chain verification")
+	}
+
+	errCh := make(chan error, 1)
+	r.actions <- func() {
+		r.syncRegister(info, errCh)
+	}
+
+	return <-errCh
+}
+
+func (r *Registry) IsRegistered(providerID peer.ID) bool {
+	done := make(chan struct{})
+	var found bool
+	r.actions <- func() {
+		_, found = r.providers[providerID]
+		close(done)
 	}
+	<-done
+	return found
 }
 
-func (r *Registry) AddProviderAddress(providerID peer.ID, addr ma.Multiaddr) bool {
-	r.rwmutex.Lock()
-	defer r.rwmutex.Unlock()
-	info, ok := r.providers[providerID]
-	if !ok {
-		r.providers[providerID] = &ProviderInfo{
-			Addresses: []ma.Multiaddr{addr},
+func (r *Registry) ProviderInfoByAddr(fcAddr string) *ProviderInfo {
+	infoChan := make(chan *ProviderInfo)
+	r.actions <- func() {
+		// TODO: consider adding a map of fcAddr->providerID
+		for _, info := range r.providers {
+			if info.FcAddress == fcAddr {
+				infoChan <- info
+				break
+			}
 		}
-		return true
+		close(infoChan)
 	}
-	for i := range info.Addresses {
-		if addr == info.Addresses[i] {
-			return false
+
+	return <-infoChan
+}
+
+func (r *Registry) ProviderInfo(providerID peer.ID) *ProviderInfo {
+	infoChan := make(chan *ProviderInfo)
+	r.actions <- func() {
+		info, ok := r.providers[providerID]
+		if ok {
+			infoChan <- info
 		}
+		close(infoChan)
 	}
-	info.Addresses = append(info.Addresses, addr)
-	return true
+
+	return <-infoChan
+}
+
+func (r *Registry) AllProviderInfo() []*ProviderInfo {
+	var infos []*ProviderInfo
+	done := make(chan struct{})
+	r.actions <- func() {
+		infos = make([]*ProviderInfo, len(r.providers))
+		var i int
+		for _, info := range r.providers {
+			infos[i] = info
+			i++
+		}
+		close(done)
+	}
+	<-done
+	return infos
+}
+
+func (r *Registry) syncStartDiscover(fcAddr string, signature, signed []byte, errCh chan<- error) {
+	err := r.syncNeedDiscover(fcAddr)
+	if err != nil {
+		r.syncEndDiscover(fcAddr, nil, err, errCh)
+		return
+	}
+
+	// Mark discovery as in progress
+	r.discoTimes[fcAddr] = time.Time{}
+
+	r.pendingDisco++
```

Comment:
i'm confused by the semantics of pendingDisco

Reply:
That is the number of discoveries that are currently in progress. This is used to determine who many to wait for to finish when doing a graceful shutdown.

Reply:
Changed to use `sync.WaitGroup` instead.  It was not strictly necessary to synchronize this since it, since incr and decr of counter was done in same goroutine, but using `sync.WaitGroup` makes the code cleaner and more obvious.